src/convex/ConvexAvatarMultiToken.sol

```diff
@@ -0,0 +1,832 @@
+    /// @dev This function doesn't claim any rewards.
+    function withdrawAll() external onlyOwnerOrManager {
+        uint256 length = baseRewardPools.length();
+        uint256[] memory curveLpDeposited = new uint256[](length);
+        for (uint256 i; i < length; i++) {
+            curveLpDeposited[i] = IBaseRewardPool(baseRewardPools.at(i)).balanceOf(address(this));
+        }
+
+        _withdraw(pids.values(), curveLpDeposited);
+    }
+
+    /// @notice Unstakes the given amount of assets and transfers them back to owner. Can only be called by owner or manager.
+    /// @dev This function doesn't claim any rewards.
+    /// @param _pids Pids targetted to withdraw from
+    /// @param _amountAssets Amount of assets to be unstaked.
+    function withdraw(uint256[] calldata _pids, uint256[] calldata _amountAssets) external onlyOwnerOrManager {
+        if (_pids.length != _amountAssets.length) {
+            revert LengthMismatch();
+        }
+        _withdraw(_pids, _amountAssets);
+    }
+
+    /// @notice Claims any pending CRV, CVX & FXS rewards and sends them to owner. Can only be called by owner.
+    /// @dev This is a failsafe to handle rewards manually in case anything goes wrong (eg. rewards need to be sold
+    ///      through other pools)
+    function claimRewardsAndSendToOwner() external onlyOwner {
+        address ownerCached = owner();
+        // 1. Claim CVX, CRV & FXS rewards
+        (uint256 totalCrv, uint256 totalCvx, uint256 totalFxs) = claimAndRegisterRewards();
+
+        // 2. Send to owner
+        CRV.safeTransfer(ownerCached, totalCrv);
+        CVX.safeTransfer(ownerCached, totalCvx);
+        if (totalFxs > 0) {
+            FXS.safeTransfer(ownerCached, totalFxs);
+        }
+    }
+
+    /// @dev given a target PID, it will add the details in the `EnumerableSet`: pids, assets & baseRewardPools
+    /// @param _newPid target pid numeric value to add in contract's storage
+    function addCurveLpPositionInfo(uint256 _newPid) external onlyOwner {
+        _addCurveLpPositionInfo(_newPid);
+    }
+
+    /// @dev given a target PID, it will remove the details from the `EnumerableSet`: pids, assets & baseRewardPools
+    /// @param _removePid target pid numeric value to remove from contract's storage
+    function removeCurveLpPositionInfo(uint256 _removePid) external onlyOwner {
+        if (!pids.contains(_removePid)) {
+            revert PidNotIncluded(_removePid);
+        }
+
+        (address lpToken,,, address crvRewards,,) = CONVEX_BOOSTER.poolInfo(_removePid);
+        IBaseRewardPool baseRewardPool = IBaseRewardPool(crvRewards);
+
+        uint256 stakedBal = baseRewardPool.balanceOf(address(this));
+        if (stakedBal > 0) {
+            revert CurveLpStillStaked(lpToken, crvRewards, stakedBal);
+        }
+
+        // NOTE: verify pending rewards and claim. Processing is done separately
+        if (baseRewardPool.earned(address(this)) > 0) {
```

Comment:
I wonder if you can getReward if rewards are zero, I think it would save gas

Reply:
looks safe to call it even when rewards are zero

https://etherscan.io/address/0x36c7E7F9031647A74687ce46A8e16BcEA84f3865#code#L980

internal call to one virtual base reward pool containing extra rewards 

https://etherscan.io/address/0x008aEa5036b819B4FEAEd10b2190FBb3954981E8#code#L874

it will not revert 👍

Reply:
resolved in f1439b3604993b3d17daaf60f95e55ad709718ef